### PR TITLE
build: add api test

### PR DESCRIPTION
**Pull request by sobolk**
*Issue #, if available:*

*Description of changes:*

This PR introduces a new mechanism to verify API surface. In a form of enforcing a test file that explores usage of symbols defined in `API.md`. Changes in these test files should then warn reviewers about breaking changes that could slip in manual `API.md` reviews.

The introduced check does the following things:
1. It asserts that every package with `API.md` file has companion `src/api.test.ts`
2. It extracts AST from `API.md` snippet and gets knowledge about symbols used there. I.e.
    1. A name of the symbol
    2. Optionally a name of parent symbol if this makes sense (for example to disambiguate prop names or parameter names should enclosing types use same names)
    3. Whether symbol is optional or not
3. It asserts that `src/api.test.ts` has two sections that explore min and max usage
4. It asserts that all recognized symbols are used in `api.test.ts`
5. It asserts that optional symbols are not used in min api usage section. (to make this section sensitive to breaking changes)

After enforcing that `src/api.test.ts` is crafted to standard above. We rely on compiler to tell us if we make breaks or not.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by edwardfoyle — COMMENTED** [review]
Overall the approach seems sound, but the code is _very_ dense. I worry about the long-term maintenance and don't want it to be more burdensome than helpful.

I wonder if we could do a hybrid between what you have here and [this approach](https://stackoverflow.com/a/71618156/5283094). We can keep the validation around min and max symbol usage, but then use that fancy type in the stackoverflow post to assert that there are no breaking changes introduced

File: packages/client-config/src/api.test.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    aws_appsync_graphqlEndpoint: '',
+    aws_appsync_region: '',
+    aws_appsync_additionalAuthenticationTypes: '',
+    aws_appsync_conflictResolutionMode: '',
+    aws_appsync_apiKey: '',
+    modelIntrospection: '',
+  };
+
+  const outDir: string | undefined = undefined;
+  const format: ClientConfigFormat | undefined = undefined;
```
[the remark this answers is not in the document]
Comment (sobolk) [reply packages/client-config/src/api.test.ts:88]: they should.
I need to think about how to add such usage assertion (and how to make adding such assertion easy).

File: scripts/check_api_tests.ts [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+      if (!isUsedInImports) {
+        const item = symbolIdentifierFactory(node as ts.Identifier);
+        if (item) {
+          accumulator.push(item);
+        }
+      }
+    }
+
+    ts.forEachChild(node, (child) =>
+      this.collectSymbolIdentifiersRecursively(
```
[the remark this answers is not in the document]
Comment (sobolk) [reply scripts/check_api_tests.ts:423]: I tried BSF with queue, but that somehow broke internal state of the tree, so I fell back to what library maintainers recommended, i.e. DFS (which I prefer with recursion than stack, but I can give it a spin if algorithm stabilizes).

File: scripts/check_api_tests.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+    node: ts.Identifier
+  ): ExpectedApiSymbol | undefined => {
+    let isOptional: boolean | undefined;
+    const name = node.getText();
+    let usagePredicate: ExpectedApiSymbolUsagePredicate | undefined;
+    let parentName: string | undefined;
+    let description: string | undefined;
+    if (
+      node.parent.kind === ts.SyntaxKind.TypeReference &&
+      ['Partial', 'Promise'].includes(name)
```
[the remark this answers is not in the document]
Comment (sobolk) [reply scripts/check_api_tests.ts:116]: type name is going to show up in error message. 
error messages should have mitigation steps (todo).

**Comment by sobolk** [comment]
> Overall the approach seems sound, but the code is _very_ dense. I worry about the long-term maintenance and don't want it to be more burdensome than helpful.
> 
> I wonder if we could do a hybrid between what you have here and [this approach](https://stackoverflow.com/a/71618156/5283094). We can keep the validation around min and max symbol usage, but then use that fancy type in the stackoverflow post to assert that there are no breaking changes introduced

If that stackoverflow works the demand for usage patterns in that test would greatly decrease (and simplify the validator).
One caveat is that we'll need to access "how test file or API.md looked like in latest release". 
Either via github lookup or by embedding artifact(s) in npm package.

**Comment by edwardfoyle** [comment]
> One caveat is that we'll need to access "how test file or API.md looked like in latest release".

Isn't it sufficient to compare to what it looks like on main? That could be done using just a bit of git commands

**Comment by sobolk** [comment]
> Actually maybe not even that. The current approach also doesn't read state from live, it just adds additional objects under test that exercise the API. Someone could make a breaking change and update the corresponding tests and everything would be green. It's just the changes to those tests signal a breaking change to the reviewer.
> 
> If we went with the other approach it would be the same thing. You create an object that conforms to the min and max usage of the API and if you have to change those objects, then it's a breaking change.

I'm starting to think that for stackoverflow approach we could just generate usage objects from API.md and let compiler do the rest. This would eliminate manual edits.

I'm running into weird corner case with functions though... See https://github.com/aws-amplify/amplify-backend/commit/1fed9dbd5abf80ed62d39b3dc9f5ad16143506ef#diff-91b4d1a523d17e351c7f279359839ad39eccd45458aac8d2df2d19abc496ef89 .

```
bcd0740fa72c:samsara-cli sobkamil$ nr build

> amplify-backend@0.0.1 build
> tsc --build packages/* && tsc --build scripts

packages/api-tests/src/client_config.ts:4:6 - error TS2300: Duplicate identifier 'NoBreakingChanges'.

4 type NoBreakingChanges = NestedAssignable<
       ~~~~~~~~~~~~~~~~~

packages/api-tests/src/client_config.ts:13:6 - error TS2300: Duplicate identifier 'NoBreakingChanges'.

13 type NoBreakingChanges = NestedAssignable<
        ~~~~~~~~~~~~~~~~~

packages/api-tests/src/client_config.ts:18:3 - error TS2344: Type 'typeof import("/Users/sobkamil/git/samsara-cli/packages/api-tests/src/client_config_new", { assert: { "resolution-mode": "import" } })' does not satisfy the constraint 'Pick<typeof import("/Users/sobkamil/git/samsara-cli/packages/api-tests/src/client_config_old", { assert: { "resolution-mode": "import" } }), "getClientConfigPathUsage">'.
  Types of property 'getClientConfigPathUsage' are incompatible.
    Type '(outDir: string, format?: ClientConfigFormat | undefined) => Promise<string>' is not assignable to type '(outDir?: string | undefined, format?: ClientConfigFormat | undefined) => Promise<string>'.
      Types of parameters 'outDir' and 'outDir' are incompatible.
        Type 'string | undefined' is not assignable to type 'string'.
          Type 'undefined' is not assignable to type 'string'.

18   typeof import('./client_config_new.js')
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 3 errors.
```

**Pull request closed without merging**